src/tests/ftest/cart/util/daos_sys_logscan.py

```diff
@@ -0,0 +1,553 @@
+        self._pools = {}
+        self._highest_pmapver = {}
+        self._cur_ldr_rank = {}
+        self._cur_ldr_pid = {}
+        self._cur_term = {}
+
+        # filename to rank map
+        self._file_to_rank = {}
+
+        self._warnings = []
+        self._check_rb_new_fmt = True
+        self._check_rb_legacy_fmt = True
+        self._debug = False
+
+    def _warn(self, wmsg, fname, line=None):
+        full_msg = f"WARN file={fname}"
+        if line:
+            full_msg += f", line={line.lineno}"
+        full_msg += f": {wmsg}"
+
+        self._warnings.append(full_msg)
+        print(full_msg)
+
+    def get_warnings(self):
+        """Return all warnings stored when scanning engine log files"""
+        return self._warnings
+
+    def _find_rank(self, log_iter):
+        print(f"INFO: searching for rank in file {log_iter.fname}")
+        found = False
+        for line in log_iter.new_iter():
+            # when a rank assignment log line found (engine start)
+            match = self.re_rank_assign.match(line.get_msg())
+            if match:
+                self._file_to_rank[log_iter.fname] = int(match.group(1))
+                found = True
+                break
+
+            # Future enhancement: what about log rotation (not an engine start scenario)?
+        return found
+
+    # return log-message, hostname, and date/timestamp components of the line
+    def _get_line_components(self, line):
+        return line.get_msg(), line.hostname, line.time_stamp
+
+    # is this rank, pid the leader of the pool with uuid puuid?
+    def _is_leader(self, puuid, rank, pid):
+        if puuid not in self._pools:
+            return False
+        if self._cur_ldr_rank[puuid] == rank and self._cur_ldr_pid[puuid] == pid:
+            return True
+        return False
+
+    def _match_ps_step_up(self, fname, line, pid, rank):
+        msg, host, datetime = self._get_line_components(line)
+        match = self.re_step_up.match(msg)
+        if match:
+            puuid = match.group(1)
+            term = int(match.group(2))
+            if puuid not in self._pools:
+                self._pools[puuid] = {}
+            self._cur_ldr_rank[puuid] = rank
+            self._cur_ldr_pid[puuid] = pid
+            self._cur_term[puuid] = term
+            old_term = term - 1
+            # if term already exists, error?
+            if term in self._pools[puuid]:
+                self._warn(f"pool {puuid} term {term} already seen!", fname, line)
+            # carry over most recent map version into the new term, avoid later KeyError
+            if old_term in self._pools:
+                if self._pools and self._pools[puuid][old_term]["maps"] != {}:
+                    last_mapver = list(self._pools[puuid][old_term]["maps"].keys())[-1]
+                    pmap_versions = self._pools[puuid][old_term]["maps"][last_mapver]
+                    pmap_versions["carryover"] = True
+                    # pmap_versions["rb_gens"] = {}
```

Comment:
Should probably be removed.
Not sure of the allowed rebuild state when a new leader is elected: Could be useful to check if it is consistent.